server/src/logic/services/filesystem-image.service.ts

```diff
@@ -0,0 +1,56 @@
+import * as fs from "fs";
+import * as winston from "winston";
+import { Config } from "./../../config";
+import { ApiError, DataType, ERRORS } from "./../../models/index";
+import { ImageService } from "./image.service";
+
+export class FilesystemImageService implements ImageService {
+
+  public persistImage (fileName: string, dataType: string, userIdentifier: string): Promise<string> {
+
+    return new Promise((resolve, reject) => {
+      const sourcePath = `${Config.IMAGE_TMP_DIR}/${fileName}`;
+      const userDataDir = `${Config.DATA_BASE_DIR}/${userIdentifier}`;
+      const dataTypeDir = `${userDataDir}/${dataType}`;
+      const targetPath = `${dataTypeDir}/${fileName}`;
+
+      const error = new ApiError(ERRORS.IMAGE_UPLOAD)
+        .addSource(FilesystemImageService.name)
+        .addSourceMethod("persistImage");
+
+      // 'this' doesn't work correctly inside of fs.readFile.
+      function createDirIfNotExists (dir: string, rej: Function): void {
```

Comment:
why have you added the function declaration into the promise? is this a common practice?